src/main/scala/asciibot/parser/parser.scala

```diff
@@ -0,0 +1,200 @@
+package asciibot.parser
+
+import java.io.BufferedReader
+
+import scala.collection.mutable.MutableList
+import picolib.maze.Maze
+import picolib.semantics._
+
+class AsciibotError() extends RuntimeException
+case class CharBlockError(m: String, sc: Int, ec: Int) extends AsciibotError
+case class BlockError(m: String, startChar: Int, endChar: Int,
+                      startLine: Int, endLine: Int) extends AsciibotError
+case class LineError(m: String, line: Int) extends AsciibotError
+case class FileError(m: String) extends AsciibotError
+
+object AsciibotParser {
+
+  def parseLines(program: Iterator[String]) : List[Rule] = {
+    val lines: MutableList[String]= new MutableList
+    val rules: MutableList[Rule] = new MutableList
+    program.zipWithIndex foreach { case (line, lineNum) => {
+      val noComment = line.split("//")(0)
+      if (noComment.trim.equals("")) {
+        if (lines.length > 0)
+          throw new LineError("Blank line in middle of rule", lineNum)
+      } else {
+        lines += noComment
+        if (lines.length == 3) {
+          try {
+            rules ++= parseLine(lines)
+          } catch {
+            case CharBlockError(m, sc, ec) => throw new BlockError(m, sc, ec, lineNum-1, lineNum+1)
+          }
+          lines.clear
+        }
+      }
+    } }
+    if (!lines.isEmpty) {
+      throw new FileError("File ended in middle of rule")
+    }
+    rules.toList
+  }
+
+  def parseLine(lns: Seq[String]) : List[Rule] = {
+    assert(lns.length == 3)
+    val maxLen = (lns map (l => l.length)).max
+    // Make sure each string is the same length
+    val nlns = lns map (_.padTo(maxLen, ' '))
+
+    val rules: MutableList[Rule] = new MutableList
+
+    var ruleStart = 0
+
+    // iterate over columns
+    var i = 0
+    for (i <- 0 until maxLen) {
+      // Number of bars in a colum
+      val numBars = (nlns map (l => l(i)) filter (_=='|')).length
+      if (!(numBars == 0 || numBars == 3)) {
+        //throw new RuleSeparatorError(i)
+        throw new CharBlockError("Unaligned vertical bars", i,i+1)
+      }
+      if (numBars == 3) {
+        rules += parseRule(nlns, ruleStart, i)
+        ruleStart = i + 1
+      }
+    }
+    rules += parseRule(nlns, ruleStart, maxLen)
+
+    rules.toList
+  }
+
+  def parseRule(lines: Seq[String], start: Int, end: Int) : Rule = {
+    def throwBE(m : String) = throw new CharBlockError(m, start, end)
+    val arrowStart = lines(1).indexOfSlice("->", start)
+    if (arrowStart < start || arrowStart >= end) {
+      throwBE("No arrow found")
+    }
+    val (inStateStart, inStateEnd) = trimBlock(lines, start, arrowStart)
+    val (outStateStart, outStateEnd) = trimBlock(lines, arrowStart + 2, end) 
+    val (surrs, inState) = parseInState(lines, inStateStart, inStateEnd)
+    val (md, outState) = parseOutState(lines, outStateStart, outStateEnd, surrs)
+    Rule(inState, surrs, md, outState)
```

Comment:
Do you do any error checking for bad states?  If you did, would you do that here or in some post-processing step?